api/bucket.go

```diff
@@ -1,19 +1,209 @@
 package api
 
 import (
-	"github.com/gin-gonic/gin"
-	"github.com/konveyor/tackle2-hub/model"
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"fmt"
 	"io"
 	"net/http"
 	"os"
+	"path"
 	pathlib "path"
+	"path/filepath"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/konveyor/tackle2-hub/model"
 )
 
 //
 // BucketHandler provides bucket management.
 type BucketHandler struct {
 }
 
+func (h *BucketHandler) serveBucketGet(ctx *gin.Context, owner *model.BucketOwner) {
+	if ctx.Request.Header.Get(Directory) == DirectoryArchive {
+		h.getDirArchive(ctx, path.Join(owner.Bucket, ctx.Param(Wildcard)))
+	} else {
+		h.content(ctx, owner)
+	}
+}
+
+func (h *BucketHandler) serveBucketUpload(ctx *gin.Context, owner *model.BucketOwner) {
+	if ctx.Request.Header.Get(Directory) == DirectoryExpand {
+		h.uploadDirArchive(ctx, path.Join(owner.Bucket, ctx.Param(Wildcard)))
+	} else {
+		h.upload(ctx, owner)
+	}
+}
+
+func (h *BucketHandler) uploadDirArchive(ctx *gin.Context, dir string) {
+	// Prepare to uncompress the uploaded data, report 4xx errors
+	file, err := ctx.FormFile("file")
+	if err != nil {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+	fileReader, err := file.Open()
+	if err != nil {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+	defer fileReader.Close()
+
+	ungzReader, err := gzip.NewReader(fileReader)
+	if err != nil {
+		ctx.Status(http.StatusBadRequest)
+		return
+	}
+	defer ungzReader.Close()
+
+	// Report 5xx errors for extraction process
+	defer func() {
+		if err != nil {
+			log.Error(err, "bucket archive expand action failed")
+			ctx.JSON(http.StatusInternalServerError, gin.H{
+				"error": err,
+			})
+			return
+		}
+	}()
+
+	// Clean and prepare destination directory
+	err = os.RemoveAll(dir)
+	if err != nil {
+		return
+	}
+	if err = os.Mkdir(dir, 0777); err != nil {
```

Comment:
I think we'll need os.MakedirAll() here to ensure intermediate directories are created.